### DISCUSSION TITLE
On redesigning Godot UI

### DISCUSSION THREAD

**Discussion by theraot** [godotengine/godot-proposals#4101] [Editor]
I believe this is a place for discussion, for not fully formed proposals. And I have some ideas about the Godot editor UI, which aren't urgent or anything like that, so I'm putting them all here for discussion purposes.

---

We have on the top of the editor "2D", "3D", "Script", and "AssetLib". They all change how the the tab that are below work. However, only "2D" and "3D" are related to the currently opened scene.

Thus I suggest the following change: Have "Script" and "AssetLib" open a new tab (or switch to if it is already open) dedicated to the scripts or the asset library respectively. These tabs would not be linked to any scene. Thus switching back to the scene could be done by clicking on the scene tab (when I'm in the script editor, and I click on a scene tab I expect to see the scene, but currently I would be still seeing the script editor).

*To be clear, I'm not suggesting that switching to the script and asset library tabs changes the currently edited scene to none. Also I'm not suggesting to change how "2D" and "3D" work. So you would still be able to switch back using those.*

---

Furthermore, the documentation opens inside the script editor. I want to suggest to open the documentation on its own tab. This change also implies adding the option to open the documentation to the top of the editor along side "2D", "3D", "Script" and "AssetLib". This way I would be able to switch from the documentation to the script by clicking on the scripts tab. Also, I think it would make documentation easier to discover.

---

Furthermore, "2D" and "3D" are about how the scene tab are presented. I don't have a better place to place them, but since they are different, a I believe a visual separation between them and the rest of options makes sense.

---

If you can take the file menu out from the script editor, and perhaps call it script and put it next to the scene menu… Then you could give each script its own tab. Like, you know, most IDEs out there. 

---

About the panels, I would like the option to have them auto-hide. Which would be conveyed with a thumbtack icon used to toggle the feature. When hidden, the panels would look like tabs on the side of the editor, and clicking on them would bring them up. Then they would hide when clicking elsewhere. And we would click on the thumbtack icon to prevent them from hiding, or to allow them to hide again.

---

If the panels can auto-hide, then the distraction free mode can go.

---

I have this 1600px by 900px monitor, which I guess is tiny for modern standards, or at least Godot makes me think that. Since the width is 1600px it means that splitting the screen between two windows requires their minimum width to be 800px, and that is more than the minimum width of Godot editor. I believe the 800px minimum width would be acceptable for everybody in distraction free mode, and I'm suggesting to get rid of distraction free mode in favor of auto-hiding panels. *<strike>Also, don't you usually have Godot maximized anyway?</strike> I don't expect changing the minimum width to bother too many people.*

**Comment by Calinou** [comment]
> Also, don't you usually have Godot maximized anyway? I don't expect changing the minimum width to bother too many people.

On larger monitors, you can tile two windows side by side. This means we shouldn't assume that the editor will always be used maximized. However, it's very unlikely you'll do this on a 1600×900 monitor – and even on 1920×1080, it's not that likely in my experience. 2560×1440 makes this kind of splitting far more usable (at 100% scaling).

**Reply by theraot** [reply]
Yeah, that makes sense. On a similar note, I would rather have some use from not having it maximized too. Currently the space left by Godot is too small for almost any other thing. So a smaller minimum a problem for a person with a bigger monitor tiling the windows as you describe? Or is there some other use case for which a smaller minimum is a problem?

**Comment by aXu-AP** [comment]
I too have thought about this 2D/3D/Script/Assets mode for as long as I have used Godot (1y). It is quite unintuitive, altough you get used to it. Assetlib would be nicer as its own tab, at it doesn't have anything to do with selected tab. However, for scripting this wouldn't be the case. You suggest that:
> To be clear, I'm not suggesting that switching to the script and asset library tabs changes the currently edited scene to none.

However, how can you tell what scene are you editing if scripting is separate tab from scenes? Notice that script editor also knows the scene that is currently open, you get suggestions for $Nodes.

For me, preferred solution would be an option to open 2D/3D editor when switching scenes. We currently have option to show script button in a tab, so I think this behaviour would go hand-in-hand with that. Want to edit visual scene? Click on tab. Want to edit script? Click on script button on tab.

**Reply by theraot** [reply]
While editing a script, there can be a scene open. As you quote me, I'm not suggesting it changes to none. The edited scene would remain the edited scene. That is not only for the node suggestion, but also to be able to see the scene and inspector panels for reference.

---

> For me, preferred solution would be an option to open 2D/3D editor when switching scenes. We currently have option to show script button in a tab, so I think this behaviour would go hand-in-hand with that. Want to edit visual scene? Click on tab. Want to edit script? Click on script button on tab.

This is good too.

**Comment by fire** [comment]
Can you provide visuals? I noticed a small amount of difficulty visualizing the proposal and I can't be the only one.

**Comment by Deozaan** [comment]
I think the whole 2D, 3D, Script, and Asset Library section needs to be rethought and possibly eliminated entirely. The asset library link doesn't belong there, IMO. You use it maybe once at the beginning of a new project and then you never need it again. And Whether it uses a 2D or 3D viewport should automatically be determined by what the root node is for the current scene. So if there was a different way that made sense to easily toggle between editing the scene and editing scripts, that section of the UI could be freed up.

> Also, don't you usually have Godot maximized anyway?

No! And I hate how Godot automatically maximizes itself every time I open it! It's one of my biggest annoyances about the engine. I really wish Godot remembered the window size and position it was last set to so I wouldn't have to reconfigure it every time I open or relaunch/reload a project.

**Reply by Calinou** [reply]
> No! And I hate how Godot automatically maximizes itself every time I open it! It's one of my biggest annoyances about the engine. I really wish Godot remembered the window size and position it was last set to so I wouldn't have to reconfigure it every time I open or relaunch/reload a project.

See https://github.com/godotengine/godot/pull/32056. In the meantime, you can write a script that uses [command line arguments](https://docs.godotengine.org/en/stable/getting_started/editor/command_line_tutorial.html) to force specific window positions and sizes.

**Comment by setanarut** [comment]
> I want to suggest to open the documentation on its own tab

+1

Also Aability to hide panels by dragging from the edge.

<img width="296" alt="Ekran Resmi 2022-02-28 00 03 51" src="https://user-images.githubusercontent.com/36481442/155899930-ad151274-8912-431b-8e6e-05e8cace5aad.png">
